Add refresh-config and open-config command

jharrilim opened the pull request:
![config-refresh](https://user-images.githubusercontent.com/19535809/158047731-e2223808-0a8a-4de8-8d71-fc0b742b520c.gif)

This PR aims to enable config refreshes while in the editor, and provide an in-editor alternative to `--edit-config`.

It adds two new `:commands`:

- `refresh-config`
- `open-config`


Basically what happens is:

1. When the editor receives the `refresh-command`, it fires off an event up to the application
2. The application sees the event and resets the config to the latest version from the config.toml file
3. The main config, editor config, and theme are refreshed

commentary:

~I couldn't get dynamic dispatch working using [DynAccess](https://docs.rs/arc-swap/latest/arc_swap/access/trait.DynAccess.html). This would presumably save a lot of work when updating the config and propagating changes.~ Figured it out, but I had to remove `#[derive(Debug)]` from `Editor`.

This is also my first time using `tokio` and `arc_swap`. I actually don't know what I'm doing, I'm being carried heavily by the compiler rn

related issues:

- https://github.com/helix-editor/helix/pull/1771
- https://github.com/helix-editor/helix/pull/1797

archseer commented on helix-term/src/application.rs:
```diff
@@ -245,6 +251,38 @@ impl Application {
         }
     }
 
+    pub fn refresh_config(&mut self) {
```
This doesn't need to be a separate event processed on the Application level, both `application.config` and `editor.config` point to the same `ArcSwap<Arc<T>>` so it's enough to add a `editor.reload_config` method that loads and updates the config via `config.store`.

archseer commented on helix-term/src/config.rs:
```diff
@@ -20,6 +21,25 @@ pub struct LspConfig {
+impl Config {
+    pub fn load(config_path: PathBuf) -> Result<Config, Error> {
+        match std::fs::read_to_string(config_path) {
+            Ok(config) => Result::Ok(toml::from_str(&config)
```
Is `Result::Ok` necessary here? `Ok` should be enough

archseer commented on helix-term/src/config.rs:
```diff
@@ -20,6 +21,25 @@ pub struct LspConfig {
+                    Config::default()
+                })),
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => Result::Ok(Config::default()),
+            Err(err) => return Err(Error::new(err)),
```
```suggestion
            Err(err) => return Err(err.into()),
```

archseer commented on helix-term/src/config.rs:
```diff
@@ -20,6 +21,25 @@ pub struct LspConfig {
+        match std::fs::read_to_string(config_path) {
+            Ok(config) => Result::Ok(toml::from_str(&config)
+                .map(merge_keys)
+                .unwrap_or_else(|err| {
+                    eprintln!("Bad config: {}", err);
+                    eprintln!("Press <ENTER> to continue with default config");
+                    use std::io::Read;
+                    // This waits for an enter press.
+                    let _ = std::io::stdin().read(&mut []);
+                    Config::default()
+                })),
```
Hmm, it seems that this case (and probably NotFound too) should be handled outside this method. This print + blocking behavior is only intended to happen during `hx` startup. You should be able to catch the error and use `editor.set_error()` when reloading

archseer commented on helix-view/src/editor.rs:
```diff
@@ -271,6 +272,8 @@ pub struct Breakpoint {
     pub log_message: Option<String>,
 }
 
+pub trait DynAccessDebug<T>: DynAccess<T> + std::fmt::Debug {}
```
?

jharrilim replied:
😳 was trying stuff out and forgot to remove that

archseer commented on helix-term/src/application.rs:
```diff
@@ -42,8 +43,7 @@ pub struct Application {
 
-    // TODO: share an ArcSwap with Editor?
-    config: Config,
+    config: Arc<ArcSwap<Config>>,
```
This seems wrong, `ArcSwap` already contains an `Arc`. 

Should be `ArcSwap<Config>`

jharrilim replied:
This pattern is [described here](https://docs.rs/arc-swap/latest/arc_swap/docs/patterns/index.html#sharing-of-configuration-data), the important bits being:

> The idea here is, each new version is a newly allocated in its own [Arc](https://doc.rust-lang.org/nightly/alloc/sync/struct.Arc.html). It is then stored into a shared ArcSwap instance.

> Note that the configuration needs to be passed through a single shared [ArcSwap](https://docs.rs/arc-swap/latest/arc_swap/type.ArcSwap.html). That means we need to share that instance and we do so through an [Arc](https://doc.rust-lang.org/nightly/alloc/sync/struct.Arc.html) (one could use a global variable instead).

I think for those reasons, we actually do want this right?

archseer replied:
Hmm, I see 👍🏻

archseer commented on helix-term/src/application.rs:
```diff
@@ -98,14 +98,16 @@ impl Application {
         });
         let syn_loader = std::sync::Arc::new(syntax::Loader::new(syn_loader_conf));
 
+        let config = Arc::new(ArcSwap::from_pointee(config));
```
```suggestion
        let config = ArcSwap::from_pointee(config);
```

archseer commented on helix-term/src/application.rs:
```diff
@@ -98,14 +98,16 @@ impl Application {
             theme_loader.clone(),
             syn_loader.clone(),
-            config.editor.clone(),
+            ArcSwap::from_pointee(config.load().editor.clone())
```
Ah right, Editor was getting a subset of the config... Let's have Editor store the whole config Arc<> as well so it shares the ArcSwap with the Application level, then add a `fn config()` getter on editor that calls `config.map(|c| c.editor)` to do a projection of that subfield https://docs.rs/arc-swap/latest/arc_swap/struct.ArcSwapAny.html#method.map

archseer replied:
I'm not sure what the lifetime on `map()` is like, we might be able to do that here and avoid the extra helper method

jharrilim replied:
Allowing the editor to store the Application config will also mean `helix-view` will have to have `helix-term` as a dependency. That isn't possible though since it will be a cyclic dependency.

I think ideally, we can follow the pattern described in the [projecting into configuration field](https://docs.rs/arc-swap/latest/arc_swap/docs/patterns/index.html#projecting-into-configuration-field) section. I couldn't quite get the code for this to work though but I can keep fiddling. I may have been half asleep when attempting it.

In the Editor's config, I tried adding:

```rust
#[derive(Debug)]
pub struct Editor {
  pub config: Box<dyn DynAccess<Config>>,
}
```

But I get an error [rustc(2077)](https://doc.rust-lang.org/error-index.html#E0277):

```
`(dyn DynAccess<Config> + 'static)` doesn't implement `std::fmt::Debug`
```

I'm not sure how to get this to play nicely with the `#[derive(Debug)]` macro.

archseer replied:
You can manually implement debug like this: https://github.com/helix-editor/helix/blob/c0dbd6dc3f147b46d4352ac78903a7a70741cfd9/helix-view/src/document.rs#L125-L148

tbh I don't find all these `Debug` implementations useful anyway (they were previously added in https://github.com/helix-editor/helix/pull/143) since both editor and doc have so many fields that you don't want to directly print them out.

Let's remove the `Debug` bound from `Editor`?

jharrilim replied:
I've removed the debug trait

archseer reviewed (changes requested):
Looks reasonable, just some cleanup:

pickfire commented on helix-term/src/commands/typed.rs:
```diff
@@ -857,28 +859,28 @@ fn setting(
-        "search.smart-case" => runtime_config.search.smart_case = arg.parse()?,
-        "search.wrap-around" => runtime_config.search.wrap_around = arg.parse()?,
-        _ => anyhow::bail!("Unknown key `{}`.", args[0]),
+    if let Ok(runtime_config) = &mut std::sync::Arc::try_unwrap(cx.editor.config.load().clone()) {
+        match key.as_ref() {
+            "scrolloff" => runtime_config.scrolloff = arg.parse()?,
+            "scroll-lines" => runtime_config.scroll_lines = arg.parse()?,
+            "mouse" => runtime_config.mouse = arg.parse()?,
+            "line-number" => runtime_config.line_number = arg.parse()?,
+            "middle-click_paste" => runtime_config.middle_click_paste = arg.parse()?,
+            "auto-pairs" => runtime_config.auto_pairs = arg.parse()?,
+            "auto-completion" => runtime_config.auto_completion = arg.parse()?,
+            "completion-trigger-len" => runtime_config.completion_trigger_len = arg.parse()?,
+            "auto-info" => runtime_config.auto_info = arg.parse()?,
+            "true-color" => runtime_config.true_color = arg.parse()?,
+            "search.smart-case" => runtime_config.search.smart_case = arg.parse()?,
+            "search.wrap-around" => runtime_config.search.wrap_around = arg.parse()?,
+            _ => anyhow::bail!("Unknown key `{}`.", args[0]),
+        }
+        cx.editor.config.store(Arc::new(runtime_config.clone()));
     }
```
Do we need an else branch for this?

pickfire commented on helix-term/src/config.rs:
```diff
@@ -20,6 +21,25 @@ pub struct LspConfig {
+impl Config {
+    pub fn load(config_path: PathBuf) -> Result<Config, Error> {
+        match std::fs::read_to_string(config_path) {
+            Ok(config) => Result::Ok(toml::from_str(&config)
```
Do we need to read from toml for config every time? Can't we just save the struct and reuse it rather than having to reparse it on every `load`?

archseer replied:
If the file changed then yes of course you need to re-read. The caching needs to happen on application level

jharrilim commented on helix-term/src/commands/typed.rs:
```diff
@@ -881,6 +879,9 @@ fn setting(
 
+    cx.editor
+        .config_events
+        .push(tokio_stream::once(ConfigEvent::Update(runtime_config)));
```
With the change to put the editor config behind dynamic dispatch, we now have to send updates back up to the application, and let it make the changes there.

archseer commented on helix-term/src/commands.rs:
```diff
@@ -848,7 +848,12 @@ fn goto_window(cx: &mut Context, align: Align) {
     // as we type
-    let scrolloff = cx.editor.config.scrolloff.min(height.saturating_sub(1) / 2);
+    let scrolloff = cx
+        .editor
+        .config
+        .load()
```
Let's add a helper `editor.config()` that  calls `self.config.load()`

archseer commented on helix-term/src/commands.rs:
```diff
@@ -1583,8 +1588,8 @@ fn rsearch(cx: &mut Context) {
     let reg = cx.register.unwrap_or('/');
-    let scrolloff = cx.editor.config.scrolloff;
-    let wrap_around = cx.editor.config.search.wrap_around;
+    let scrolloff = cx.editor.config.load().scrolloff;
+    let wrap_around = cx.editor.config.load().search.wrap_around;
 
```
Load is an atomic operation so you only want to do it once in each method:

```
let config = cx.editor.config();

.. use config
```

archseer commented on helix-view/src/editor.rs:
```diff
@@ -305,6 +305,14 @@ pub struct Editor {
 
     pub exit_code: i32,
+
+    pub config_events: SelectAll<tokio_stream::Once<ConfigEvent>>,
```
Let's instead use `tokio::sync::mpsc::unbounded_channel`, no need for a `SelectAll`

archseer commented on book/src/generated/typable-cmd.md:
```diff
@@ -53,3 +53,5 @@
 | `:sort` | Sort ranges in selection. |
 | `:rsort` | Sort ranges in selection in reverse order. |
 | `:tree-sitter-subtree`, `:ts-subtree` | Display tree sitter subtree under cursor, primarily for debugging queries. |
+| `:refresh-config` | Refreshes helix's config. |
```
Let's call it `:reload-config`. Better yet, maybe `:config-reload`, `:config-open` so they're grouped in the list

archseer reviewed (changes requested):
Sorry it took me a few days for another look, was trying to think of a way to avoid that extra channel in the event loop but I think this is the simplest implementation for now.

Just some minor changes

archseer commented:
(By the way, feel free to join us on the Matrix channel!)

jharrilim commented:
I've made all requested changes:

- commands renamed to `:config-reload` and `:config-save`
- `config()` helper created
- keymaps now have dynamic access, tested that `:config-reload` works with this now
- `config_events` uses an unbounded channel instead of `SelectAll`
- The other refactors

And also:

- Removed `--edit-config`
- Added `keymaps()` helper
- Refactored `handle_keymap_event` (this was necessary due to `mut` issues occuring after changing `keymaps`)

archseer commented on helix-term/src/ui/editor.rs:
```diff
@@ -1182,12 +1183,19 @@ impl Component for EditorView {
-                                _ => unimplemented!(),
-                            };
+                        self.last_insert.0 = match self
+                            .keymaps
+                            .load()
```
`keymaps()`?

archseer commented on helix-term/src/ui/editor.rs:
```diff
@@ -1182,12 +1183,19 @@ impl Component for EditorView {
+                        self.last_insert.0 = match self
+                            .keymaps
+                            .load()
+                            .clone()
+                            .get_mut(&mode)
```
I'm not clear on why this was doing `get_mut()`, you could probably avoid the `.clone()` by using `.get()`

jharrilim replied:
From a glance, it looks like it was done because of the `unwrap()` after it

archseer commented on helix-term/src/application.rs:
```diff
@@ -98,30 +101,34 @@ impl Application {
             syn_loader.clone(),
-            config.editor.clone(),
+            Box::new(Map::new(Arc::clone(&config), |config: &Config| {
+                &config.editor
+            })),
         );
 
-        let editor_view = Box::new(ui::EditorView::new(std::mem::take(&mut config.keys)));
+        let keymaps = Box::new(Map::new(Arc::clone(&config), |config: &Config| {
+            &config.keys
+        }));
```
Can we avoid boxing by using `Map<>` directly? For that to work you will probably need to use a `fn` as the last parameter rather than a closure, so that you can name it on the type

archseer commented on helix-term/src/ui/editor.rs:
```diff
@@ -698,51 +697,51 @@ impl EditorView {
-                    if let Some(ch) = event.char() {
-                        commands::insert::insert_char(cx, ch)
-                    }
+        let mut keymaps = self.keymaps().clone();
+        let insert_keys = keymaps.get_mut(&Mode::Insert).unwrap();
```
Hmm, this is tricky because some of these commands expect to modify the `keymaps.state`. If you `load()`/`clone()` the entire keymap, then this is applied to a separate instance of the keymap, losing the persistent changes.

You want to only use ArcSwap for the `root` field inside Keymaps: https://github.com/helix-editor/helix/blob/e6a2a605495365b1ac10c1a4f94506de39b70689/helix-term/src/keymap.rs#L331-L334

archseer replied:
Doing this will also avoid the need to use atomic operations for `pending()` etc., cases where you don't need to interact with the trie

jharrilim replied:
I went off the deep end trying to get this to work, and it seems pretty hard at the moment.

It kind of went like:

- Change root to `ArcSwap<KeyTrie>`
  - `PartialEq` fails, change `root` to `KeyTrieRoot(ArcSwap<KeyTrie>)` and implement `PartialEq`
- Had no way to update the `Keymaps` itself anymore
  - Tried out sending an event back up to the event loop for this too
    - Could not re-use the same `ConfigEvent` channel since `Keymaps` isn't in the same crate as `ConfigEvent`, had to make a new channel
    - _I think_ this caused me to start re-writing stuff in a more FP approach internally in Keymaps
      - This introduced bugs when merging keys and I couldn't figure out why

It might be easier to just put `Keymaps` inside an ArcSwap and manually sync it

archseer replied:
Sorry, I led you down the wrong path.. I thought `Keymap` was the toplevel type, but it turns out to be `Keymaps`.

It seems weird to me that each `Keymap` has a `state` and `sticky` node, I'd move those to the top-level `Keymaps` struct, then you should be able to just swap the hashmap part of `Keymaps`. I can give it a go at solving this?

archseer replied:
I implemented the changes so it should be easier to make `keymaps.map` a `DynAccess` field :)

archseer replied:
Sorry actually, I ended up rebasing the commits and pushing them to master. Can you rebase?

archseer reviewed (changes requested):
Getting closer, some thoughts about the keymaps type

jharrilim commented on helix-term/src/keymap.rs:
```diff
@@ -468,169 +472,222 @@ impl Keymaps {
-                "a" => goto_prev_parameter,
-                "o" => goto_prev_comment,
-                "space" => add_newline_above,
+fn default_keymaps() -> HashMap<Mode, Keymap> {
```
(1/2) Had to move the default keymaps into its own function, this was to allow...

jharrilim commented on helix-term/src/keymap.rs:
```diff
@@ -647,194 +704,153 @@ impl Default for Keymaps {
-    let mut delta = std::mem::take(&mut config.keys);
-    for (mode, keys) in &mut config.keys.map {
-        keys.merge(delta.map.remove(mode).unwrap_or_default())
+    let mut delta = std::mem::replace(&mut config.keys, default_keymaps());
```
(2/2) ...this merge to continue working

jharrilim commented on helix-term/src/config.rs:
```diff
@@ -12,7 +14,7 @@ pub struct Config {
     pub lsp: LspConfig,
     #[serde(default)]
-    pub keys: Keymaps,
+    pub keys: HashMap<Mode, Keymap>,
```
The "Keymaps" struct is now divorced from the config. Since the `Keymaps` encapsulated view state alongside config, it wasn't quite fitting, and had to be changed for config projection to work accordingly.

jharrilim commented on helix-term/src/keymap.rs:
```diff
@@ -381,30 +386,28 @@ impl Default for Keymap {
 
 impl Keymaps {
-    pub fn new(map: HashMap<Mode, Keymap>) -> Self {
+    pub fn new(map: Box<dyn DynAccess<HashMap<Mode, Keymap>>>) -> Self {
```
I kept this as `Box<dyn DynAccess<T>>` because defining a projection function here would prevent using `arc_swap::access::Constant` which is useful for testing.

jharrilim reviewed (commented):
@archseer made changes and actually got it working! I left some comments inline describing some of the changes.

jharrilim commented on helix-term/src/keymap/default_keymaps.rs:
```diff
@@ -0,0 +1,359 @@
+use super::macros::keymap;
+use super::{Mode, Keymap};
+
+pub fn default_keymaps() -> HashMap<Mode, Keymap> {
```
I split this into its own file, as it is large enough by itself and is worth having its own point of reference.

archseer replied:
Let's name the file `default.rs` and `fn default` since `keymap` is already in the module name. This will impact some of the open PRs though

jharrilim replied:
renamed!

archseer commented on helix-term/src/keymap.rs:
```diff
@@ -419,7 +305,8 @@ impl Keymaps {
     pub fn get(&mut self, mode: Mode, key: KeyEvent) -> KeymapResult {
         // TODO: remove the sticky part and look up manually
-        let keymap = &self.map[&mode];
+        let keymaps = self.map().clone();
```
Why the need to clone here? I imagine this is cloning the entire keymap which might be very heavy.

jharrilim replied:
Good call, I thought this was an arc clone but it was not. I can change this to `&*self.map()`

archseer commented on helix-term/src/keymap.rs:
```diff
@@ -470,372 +357,32 @@ impl Keymaps {
-            Mode::Select => Keymap::new(select),
-            Mode::Insert => Keymap::new(insert),
-        ))
+impl PartialEq for Keymaps {
+    fn eq(&self, other: &Self) -> bool {
+        let a = *self.map() == *other.map();
+        a && self.state == other.state && self.sticky == other.sticky
```
Is PartialEq still required? I'd remove it from the type

jharrilim replied:
removed it!

archseer commented on helix-term/src/application.rs:
```diff
@@ -245,6 +257,52 @@ impl Application {
+            ConfigEvent::Update(editor_config) => {
+                let mut app_config = (*self.config.load().clone()).clone();
+                app_config.editor = editor_config;
+                self.config.swap(Arc::new(app_config));
```
^ this is still using `swap()`

jharrilim replied:
Ah, regression when I had reverted

archseer commented on helix-term/src/keymap/macros.rs:
```diff
@@ -0,0 +1,127 @@
+    };
+}
+
+pub use alt;
+pub use ctrl;
+pub use key;
+pub use keymap;
+pub use shift;
```
I don't think this is necessary? `macro_export` should be enough

jharrilim replied:
This caught me off guard as well, `macro_export` is a bit different from `macro_use`:
https://stackoverflow.com/a/31749071/11886888

Without the `pub use`, it can't be imported from other files

archseer reviewed (approved):
Tested and works great! I'm quite happy with the keymap refactor too, thanks for working on this! 🎉
